File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/auth_verification.ts
/**
 * Auth verification dispatcher for iOS and Android attestation
 */

import { verifyiOSAuth, type AppAttestVerdict } from './ios';
import { verifyAndroidAuth, type PlayIntegrityVerdict } from './android';
import { trackEvent } from './logging';

/**
 * Verification result interface
 */
export interface AuthVerificationResult {
  verified: boolean;
  platform: string;
  message: string;
  timestamp: string;
  challengeHash?: string;
  clientId?: string;
  attestJsonLength?: number;
  chainLength?: number;
  rootCA?: string;
  integrityVerdict?: PlayIntegrityVerdict;
  appAttestVerdict?: AppAttestVerdict;
  leafCertValidityWarning?: string;
  warnings?: string[];
}

/**
 * Message data interface for attestation verification
 */
export interface AttestationMessageData {
  challengeHash?: string;
  clientId: string;
  system: string;
  publicCert?: string; // PEM-encoded leaf certificate
}

/**
 * Verify attestation JSON based on the platform system
 * @param messageData - Parsed message object containing challengeHash, clientId, system, publicCert
 * @param attestJson - The attestation JSON from the request
 * @returns Verification result object
 */
export async function verifyAuthBySystem(
  messageData: AttestationMessageData,
  attestJson: string
): Promise<AuthVerificationResult> {
  const systemLower = messageData.system.toLowerCase();

  trackEvent('AuthVerification.Dispatch', {
    platform: systemLower,
    clientId: messageData.clientId,
    attestJsonLength: attestJson.length,
  });

  if (systemLower === 'ios') {
    return await verifyiOSAuth(messageData, attestJson);
  } else if (systemLower === 'android') {
    return await verifyAndroidAuth(messageData, attestJson);
  } else {
    // This should never happen due to validation in the route handler
    trackEvent('AuthVerification.UnsupportedSystem', {
      platform: messageData.system,
      clientId: messageData.clientId,
    });
    return {
      verified: false,
      platform: 'unknown',
      message: `Unsupported system: ${messageData.system}`,
      timestamp: new Date().toISOString(),
    };
  }
}

File: backend_samples/library/python/AzureAIVisionFaceDeviceAttestation/azure_ai_vision_face_deviceattestation/auth_verification.py
"""
Auth verification dispatcher for iOS and Android attestation.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any

from .logging import track_event
from .android import verify_android_auth
from .ios import verify_ios_auth


async def verify_auth_by_system(message_data: dict, attest_json: str) -> dict:
    system_lower = (message_data.get("system") or "").lower()
    track_event(
        "AuthVerification.Dispatch",
        {
            "platform": system_lower,
            "clientId": message_data.get("clientId"),
            "attestJsonLength": len(attest_json),
        },
    )

    if system_lower == "ios":
        return await verify_ios_auth(message_data, attest_json)
    if system_lower == "android":
        return await verify_android_auth(message_data, attest_json)

    track_event(
        "AuthVerification.UnsupportedSystem",
        {"platform": message_data.get("system"), "clientId": message_data.get("clientId")},
    )
    return {
        "verified": False,
        "platform": "unknown",
        "message": f"Unsupported system: {message_data.get('system')}",
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
